lib/utils/resolveObject.js

```diff
@@ -150,24 +150,24 @@ function selfRef(str, obj) {
       context = current_context.join(options.separator);
 
   for (i = 0; i < array.length; i++) {
-    if (ref[array[i]]) {
+    // Check for undefined as 0 is a valid, truthy value
+    if (typeof ref[array[i]] !== 'undefined') {
       ref = ref[array[i]];
     } else {
-      ref = obj;
+      // set the reference as undefined if we don't find anything
+      ref = undefined;
       break;
     }
   }
 
-  // If the reference doesn't change then it means
-  // we didn't find it in the object
-  if (ref === obj) {
+  if (typeof ref === 'undefined') {
     GroupMessages.add(
       PROPERTY_REFERENCE_WARNINGS,
       "Reference doesn't exist: " + context + " tries to reference " + str + ", which is not defined"
     );
+  } else {
+    return ref;
   }
-
-  return ref;
```

Comment:
It looks like in the previous version a bad ref would result in this function returning "obj" (ref = obj on line 156, return ref here). Now it doesn't return anything if there's a bad ref. Don't know if this is good or bad, just wondering if it might have the potential to break something else.

Reply:
Oh good point I forgot to mention that part. So I don't know why we were returning the whole object, if you look at line 91 we were testing for truthiness, which if we passed an object back would always be true. Then in line 131 we set `to_ret = ref` which would be that object as well and return it. This is inside a string.replace function, and returning an object does nothing to the string. So this was always "working", but not how we thought it was...

Reply:
And hopefully all our glorious unit tests would catch anything major, but seems everything is in order.

Reply:
This file is so hard to understand - I'm concerned about the out-of-scope variable (updated_object) used within. I'm pretty sure I added the code that sets ref = obj as part of the work to provide better error output (although blame does not seem to think so), but I'm not clear how it interfaces with the updated_object variable.

Reply:
ok so updated_object is created immediate upon calling resolveObject. It is a deep clone of the object being resolved, to prevent any changes that are made during this process from trickling back outside the resolveObject scope. I think everything is ok as is...